[GTK] Correct scaling for DPI.

gwhitney opened the pull request:
#### 69ad530f10603697a1c53b063840abc8ba7d9efe
<pre>
[GTK] Correct scaling for DPI.
<a href="https://bugs.webkit.org/show_bug.cgi?id=247980">https://bugs.webkit.org/show_bug.cgi?id=247980</a>

Reviewed by NOBODY (OOPS!).

Move the scaling factor formerly in WebKitWebView.cpp (responsive to
the xft-dpi setting) into WebKitWebViewBase.cpp, and split it into two
scaling factors. The first, a page scaling factor, uses screenDPI to
ensure that a length specified by a CSS unit of 1in will appear on the
screen as one physical inch. The second, a text scaling factor, ensures
that text specified as 96px in size will have a height on-screen as
specified by fontDPI (which is primarily driven by the xft-dpi setting),
in terms of the number of screen pixels taken up.

The two scaling factors are refreshed when any one of the inputs for
computing them changes: the GDK_SCALE factor, the xft-dpi setting, or
the screenDPI of the device it is displayed on (say by being dragged
to a different monitor). Throughout, at the default zoom level, a CSS
1in dimension is kept at one physical inch.

The scale factors are kept internal to WebKitWebViewBase (which helps
keep GTK dependencies grouped and reduces the total code in WebKitWebView
enclosed in `#ifdef PLATFORM(GTK)`) but made available to WebKitWebView
so that its external interfaces can maintain zoom level 1 as the default
zoom level, and zoom in and out from there.

So indeed, if a user of WebKitWebView asks for a zoom level of 1.5, a
length specified in CSS as 1in will then occupy one and a half physical
inches on screen.

* Source/WebCore/platform/gtk/PlatformScreenGtk.cpp:
(WebCore::fontDPI):
* Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
(webkitWebViewConstructed):
(webkitWebViewDispose):
(webkit_web_view_set_zoom_level):
(webkit_web_view_get_zoom_level):
* Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp:
(_WebKitWebViewBasePrivate::_WebKitWebViewBasePrivate):
(refreshInternalScaling):
(webkitWebViewBaseUpdateDisplayID):
(webkitWebViewBaseDispose):
(webkitWebViewBaseGetScaleFactors):
(deviceScaleFactorChanged):
(webkitWebViewBaseCreateWebPage):
* Source/WebKit/UIProcess/API/gtk/WebKitWebViewBasePrivate.h:
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/69ad530f10603697a1c53b063840abc8ba7d9efe

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/49580 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/28865 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/52627 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/52822 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/256 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/51884 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/34885 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/26485 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/40474 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/51680 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/26400 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/42706 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/21589 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/23853 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/43888 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/7951 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/45771 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/44394 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/54399 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/24665 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/20846 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/47845 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/25934 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/42850 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/46869 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10881 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/26778 "Built successfully") | | | 
| | [❌ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/25658 "Failed to checkout and rebase branch from PR 26938") | | | 
<!--EWS-Status-Bubble-End-->

mcatanzaro commented on Source/WebCore/platform/gtk/PlatformScreenGtk.cpp:
```diff
@@ -86,6 +86,16 @@ bool screenHasInvertedColors()
+    // additional font scaling factor.
+    auto* display = gdk_display_get_default();
+    if (display) {
+        if (auto* screen = gdk_display_get_default_screen(display))
```
Style nit:

```suggestion
    if (auto* display = gdk_display_get_default()) {
        if (auto* screen = gdk_display_get_default_screen(display))
```

gwhitney replied:
Changed as requested.

mcatanzaro commented on Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp:
```diff
@@ -339,6 +342,8 @@ struct _WebKitWebViewBasePrivate {
 
     PlatformDisplayID displayID;
+    double pageScaleFactor; // Corrects length of absolute units
+    double textScaleFactor; // Respects GTK font-specific scaling
```
Any idea what we should do with https://bugs.webkit.org/show_bug.cgi?id=250138, which complains that the text scale factor is used only for text and not also for the rest of the web content?

gwhitney replied:
Ah, I hadn't been aware of those reports. Give me a couple of days to reproduce them and see how this change affects them. Ultimately it should always be possible to get decent-looking results with this PR, but yes, it does give two "knobs" because GTK is set up that way with two notions of scaling, and the font scaling, if I am understanding GTK documentation and the behavior of other GTK applications correctly, _is_ intended only to scale fonts, leaving layouts and geometry unaffected. (There are lots of other GTK applications besides those built with WebKitGTK that suffer from text being drawn outside the space allocated for it with some combinations of screen and text scaling, believe me!) But I would like to make a coherent response to 250138; I will likely post something there referring to this PR. In the meantime, would you like me to mark this PR as draft?

mcatanzaro replied:
I'll let you decide. I'm fine with landing this now and leaving 250138 open for follow-up. Maybe it's better to wait a few days for you to investigate, though. Thanks for your help with this regardless.

gwhitney replied:
OK, marking as draft for now then.

mcatanzaro commented on Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp:
```diff
@@ -2555,9 +2601,27 @@ WebPageProxy* webkitWebViewBaseGetPage(WebKitWebViewBase* webkitWebViewBase)
+    // reliably. Maybe there is some kind of race condition with whatever
+    // updates the screen DPI. So instead, update the scaling directly.
+    if (std::abs(scaleRatio - 1) > 0.01) {
+        page->setPageZoomFactor(page->pageZoomFactor()*scaleRatio);
```
```suggestion
        page->setPageZoomFactor(page->pageZoomFactor() * scaleRatio);
```

gwhitney replied:
Spaces inserted.

mcatanzaro commented on Source/WebCore/platform/gtk/PlatformScreenGtk.cpp:
```diff
@@ -86,6 +86,16 @@ bool screenHasInvertedColors()
+            return gdk_screen_get_resolution(screen);
+    }
+#endif
     static GtkSettings* gtkSettings = gtk_settings_get_default();
```
```suggestion
#endif

    static GtkSettings* gtkSettings = gtk_settings_get_default();
```

gwhitney replied:
Added blank line as requested.

mcatanzaro reviewed (commented):
Looks fine to me. Just a few nits regarding style.

mcatanzaro reviewed (approved):
Thanks for your comment in https://bugs.webkit.org/show_bug.cgi?id=250138. I'm going to click the Approve button. Time to update with my suggested changes and mark as ready rather than Draft.

mcatanzaro reviewed (changes requested):
Unfortunately this breaks the Large Text accessibility setting, see https://bugs.webkit.org/show_bug.cgi?id=250138. We'll need to figure this out before landing.

mcatanzaro reviewed (approved):
Whoops, sorry... I failed to test whether it worked _without_ changes, and it doesn't. Seems the large text setting is ~~only respected when running in flatpak~~ not respected if I fail to change it properly. Let's not block on this.

gwhitney committed:
[GTK] Correct scaling for DPI.
https://bugs.webkit.org/show_bug.cgi?id=247980

Reviewed by Michael Catanzaro.

Move the scaling factor formerly in WebKitWebView.cpp (responsive to
the xft-dpi setting) into WebKitWebViewBase.cpp, and split it into two
scaling factors. The first, a page scaling factor, uses screenDPI to
ensure that a length specified by a CSS unit of 1in will appear on the
screen as one physical inch. The second, a text scaling factor, ensures
that text specified as 96px in size will have a height on-screen as
specified by fontDPI (which is primarily driven by the xft-dpi setting),
in terms of the number of screen pixels taken up.

The two scaling factors are refreshed when any one of the inputs for
computing them changes: the GDK_SCALE factor, the xft-dpi setting, or
the screenDPI of the device it is displayed on (say by being dragged
to a different monitor). Throughout, at the default zoom level, a CSS
1in dimension is kept at one physical inch.

The scale factors are kept internal to WebKitWebViewBase (which helps
keep GTK dependencies grouped and reduces the total code in WebKitWebView
enclosed in `#ifdef PLATFORM(GTK)`) but made available to WebKitWebView
so that its external interfaces can maintain zoom level 1 as the default
zoom level, and zoom in and out from there.

So indeed, if a user of WebKitWebView asks for a zoom level of 1.5, a
length specified in CSS as 1in will then occupy one and a half physical
inches on screen.

* Source/WebCore/platform/gtk/PlatformScreenGtk.cpp:
(WebCore::fontDPI):
* Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
(webkitWebViewConstructed):
(webkitWebViewDispose):
(webkit_web_view_set_zoom_level):
(webkit_web_view_get_zoom_level):
* Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp:
(_WebKitWebViewBasePrivate::_WebKitWebViewBasePrivate):
(refreshInternalScaling):
(webkitWebViewBaseUpdateDisplayID):
(webkitWebViewBaseDispose):
(webkitWebViewBaseGetScaleFactors):
(deviceScaleFactorChanged):
(webkitWebViewBaseCreateWebPage):
* Source/WebKit/UIProcess/API/gtk/WebKitWebViewBasePrivate.h:

Canonical link: https://commits.webkit.org/278128@main
